### ISSUE TITLE
Unable to install docker docker-ce-17.03.1.ce-1.el7.centos

### ISSUE THREAD

**Issue by the-nw1-group** [moby/moby#33930]
**Description**

I'm unable to install Docker CE version 17.03.1.ce-1.el7.centos (or 17.03.2.ce-1.el7.centos), on a fresh installation of Centos 7.3

**Steps to reproduce the issue:**
1. sudo yum-config-manager -y --add-repo https://download.docker.com/linux/centos/docker-ce.repo
2. sudo yum install -y docker-ce-17.03.1.ce-1.el7.centos

**Describe the results you received:**
```
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be installed
--> Processing Dependency: docker-ce-selinux >= 17.03.1.ce-1.el7.centos for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
Package docker-ce-selinux is obsoleted by docker-ce, but obsoleting package does not provide for requirements
--> Processing Dependency: libcgroup for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
--> Processing Dependency: libseccomp.so.2()(64bit) for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
--> Processing Dependency: libltdl.so.7()(64bit) for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be installed
--> Processing Dependency: docker-ce-selinux >= 17.03.1.ce-1.el7.centos for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
Package docker-ce-selinux is obsoleted by docker-ce, but obsoleting package does not provide for requirements
---> Package libcgroup.x86_64 0:0.41-11.el7 will be installed
---> Package libseccomp.x86_64 0:2.3.1-2.el7 will be installed
---> Package libtool-ltdl.x86_64 0:2.4.2-22.el7_3 will be installed
--> Finished Dependency Resolution
Error: Package: docker-ce-17.03.1.ce-1.el7.centos.x86_64 (docker-ce-stable)
           Requires: docker-ce-selinux >= 17.03.1.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.0.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.0.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.1.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.1.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.2.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.2.ce-1.el7.centos
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```
**Describe the results you expected:**

For Docker to be installed

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Azure, Centos 7.3 OpenLogic image

Running either of the following also failed in a similar way:
```
sudo yum install -y docker-ce-17.03.1.ce-1.el7.centos docker-ce-selinux.17.03.1.ce-1.el7.centos
sudo yum install -y docker-ce-17.03.2.ce-1.el7.centos docker-ce-selinux.17.03.2.ce-1.el7.centos
```
But running the following does work, but tries to install version 17.06.0.ce-1.el7.centos:
```
sudo yum install  docker-ce
```

**Comment by cpuguy83** [comment]
You must install both packages at the same time: `yum install docker-ce-17.03.1.ce-1.el7.centos docker-ce-selinux-17.03.1.ce-1.el7.centos`

**Comment by robertpearce** [comment]
nope, just got here form running into the same issue myself.

$ sudo yum install docker-ce-17.03.1.ce-1.el7.centos docker-ce-selinux-17.03.1.ce-1.el7.centos
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: centos.mirror.digitalpacific.com.au
 * epel: epel.mirror.digitalpacific.com.au
 * extras: centos.mirror.digitalpacific.com.au
 * updates: mirror.nsw.coloau.com.au
Package docker-ce-selinux is obsoleted by docker-ce, trying to install docker-ce-17.06.0.ce-1.el7.centos.x86_64 instead
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be installed
--> Processing Dependency: docker-ce-selinux >= 17.03.1.ce-1.el7.centos for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
Package docker-ce-selinux is obsoleted by docker-ce, but obsoleting package does not provide for requirements
---> Package docker-ce.x86_64 0:17.06.0.ce-1.el7.centos will be installed
--> Processing Dependency: container-selinux >= 2.9 for package: docker-ce-17.06.0.ce-1.el7.centos.x86_64
--> Running transaction check
---> Package container-selinux.noarch 2:2.19-2.1.el7 will be installed
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be installed
--> Processing Dependency: docker-ce-selinux >= 17.03.1.ce-1.el7.centos for package: docker-ce-17.03.1.ce-1.el7.centos.x86_64
Package docker-ce-selinux is obsoleted by docker-ce, but obsoleting package does not provide for requirements
--> Finished Dependency Resolution
Error: Package: docker-ce-17.03.1.ce-1.el7.centos.x86_64 (docker-ce-stable)
           Requires: docker-ce-selinux >= 17.03.1.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.0.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.0.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.1.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.1.ce-1.el7.centos
           Available: docker-ce-selinux-17.03.2.ce-1.el7.centos.noarch (docker-ce-stable)
               docker-ce-selinux = 17.03.2.ce-1.el7.centos
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest

**Comment by andrewhsu** [comment]
A new obsoletes restriction was introduced in docker-ce 17.06.0 and it looks like the yum repo is applying it to all versions of the docker-ce indiscriminately. To work around this when installing older versions of docker-ce, you can pass a flag to ignore obsoletes:
```
$ yum install -y --setopt=obsoletes=0 \
  docker-ce-17.03.1.ce-1.el7.centos \
  docker-ce-selinux-17.03.1.ce-1.el7.centos
```

Related to issue: https://github.com/docker/for-linux/issues/20#issuecomment-312122325

**Comment by the-nw1-group** [comment]
Thanks @andrewhsu that worked a treat.

**Comment by shekarvt** [comment]
thanks @andrewhsu it worked as below 
 $ yum install -y --setopt=obsoletes=0 \
  docker-ce-17.03.0.ce-1.el7.centos \
  docker-ce-selinux-17.03.0.ce-1.el7.centos

**Comment by rickbradshaw** [comment]
I see that the latest version that is actually installable is 17.05. Anything after this version has a failure on the container-selinux package which I can't find in any of my repositories that are configured. 

`root@13d2ce1ecaa9: sudo yum list docker-ce
Loaded plugins: fastestmirror, versionlock
Loading mirror speeds from cached hostfile
Installed Packages
docker-ce.x86_64                                              17.03.1.ce-1.el7.centos                                               @docker-ce-stable
Available Packages
docker-ce.x86_64                                              17.03.0.ce-1.el7.centos                                               docker-ce-stable 
docker-ce.x86_64                                              17.03.1.ce-1.el7.centos                                               docker-ce-stable 
docker-ce.x86_64                                              17.03.2.ce-1.el7.centos                                               docker-ce-stable 
docker-ce.x86_64                                              17.04.0.ce-1.el7.centos                                               docker-ce-edge   
docker-ce.x86_64                                              17.05.0.ce-1.el7.centos                                               docker-ce-edge   
docker-ce.x86_64                                              17.06.0.ce-1.el7.centos                                               docker-ce-edge   
docker-ce.x86_64                                              17.06.0.ce-1.el7.centos                                               docker-ce-stable 
docker-ce.x86_64                                              17.06.1.ce-1.el7.centos                                               docker-ce-edge   
docker-ce.x86_64                                              17.06.1.ce-1.el7.centos                                               docker-ce-stable 
docker-ce.x86_64                                              17.07.0.ce-1.el7.centos                                               docker-ce-edge   
root@13d2ce1ecaa9: yum install -y --setopt=obsoletes=0   docker-ce-17.07.0.ce-1.el7.centos   docker-ce-selinux-17.07.0.ce-1.el7.centos
Loaded plugins: fastestmirror, versionlock
Loading mirror speeds from cached hostfile
No package docker-ce-selinux-17.07.0.ce-1.el7.centos available.
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be updated
---> Package docker-ce.x86_64 0:17.07.0.ce-1.el7.centos will be obsoleting
--> Processing Dependency: container-selinux >= 2.9 for package: docker-ce-17.07.0.ce-1.el7.centos.x86_64
---> Package docker-ce-selinux.noarch 0:17.03.1.ce-1.el7.centos will be obsoleted
--> Finished Dependency Resolution
Error: Package: docker-ce-17.07.0.ce-1.el7.centos.x86_64 (docker-ce-edge)
           Requires: container-selinux >= 2.9
 You could try using --skip-broken to work around the problem

root@13d2ce1ecaa9: yum install -y --setopt=obsoletes=0   docker-ce-17.06.1.ce-1.el7.centos   docker-ce-selinux-17.06.1.ce-1.el7.centos
Loaded plugins: fastestmirror, versionlock
Loading mirror speeds from cached hostfile
No package docker-ce-selinux-17.06.1.ce-1.el7.centos available.
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be updated
---> Package docker-ce.x86_64 0:17.06.1.ce-1.el7.centos will be obsoleting
--> Processing Dependency: container-selinux >= 2.9 for package: docker-ce-17.06.1.ce-1.el7.centos.x86_64
---> Package docker-ce-selinux.noarch 0:17.03.1.ce-1.el7.centos will be obsoleted
--> Finished Dependency Resolution
Error: Package: docker-ce-17.06.1.ce-1.el7.centos.x86_64 (docker-ce-edge)
           Requires: container-selinux >= 2.9
 You could try using --skip-broken to work around the problem

root@13d2ce1ecaa9:~: yum install -y --setopt=obsoletes=0   docker-ce-17.06.0.ce-1.el7.centos   docker-ce-selinux-17.06.0.ce-1.el7.centos
Loaded plugins: fastestmirror, versionlock
Loading mirror speeds from cached hostfile
No package docker-ce-selinux-17.06.0.ce-1.el7.centos available.
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be updated
---> Package docker-ce.x86_64 0:17.06.0.ce-1.el7.centos will be obsoleting
--> Processing Dependency: container-selinux >= 2.9 for package: docker-ce-17.06.0.ce-1.el7.centos.x86_64
---> Package docker-ce-selinux.noarch 0:17.03.1.ce-1.el7.centos will be obsoleted
--> Finished Dependency Resolution
Error: Package: docker-ce-17.06.0.ce-1.el7.centos.x86_64 (docker-ce-edge)
           Requires: container-selinux >= 2.9
 You could try using --skip-broken to work around the problem
root@13d2ce1ecaa9: yum install -y --setopt=obsoletes=0   docker-ce-17.05.0.ce-1.el7.centos   docker-ce-selinux-17.05.0.ce-1.el7.centos
Loaded plugins: fastestmirror, versionlock
Loading mirror speeds from cached hostfile
Resolving Dependencies
--> Running transaction check
---> Package docker-ce.x86_64 0:17.03.1.ce-1.el7.centos will be updated
---> Package docker-ce.x86_64 0:17.05.0.ce-1.el7.centos will be an update
---> Package docker-ce-selinux.noarch 0:17.03.1.ce-1.el7.centos will be updated
---> Package docker-ce-selinux.noarch 0:17.05.0.ce-1.el7.centos will be an update
--> Finished Dependency Resolution

Dependencies Resolved

=====================================================================================================================================================
 Package                              Arch                      Version                                      Repository                         Size
=====================================================================================================================================================
Updating:
 docker-ce                            x86_64                    17.05.0.ce-1.el7.centos                      docker-ce-edge                     19 M
 docker-ce-selinux                    noarch                    17.05.0.ce-1.el7.centos                      docker-ce-edge                     28 k

Transaction Summary
=====================================================================================================================================================
Upgrade  2 Packages

Total download size: 20 M
Downloading packages:
Delta RPMs disabled because /usr/bin/applydeltarpm not installed.
(1/2): docker-ce-selinux-17.05.0.ce-1.el7.centos.noarch.rpm                                                                   |  28 kB  00:00:00     
(2/2): docker-ce-17.05.0.ce-1.el7.centos.x86_64.rpm                                                                           |  19 MB  00:00:32     
-----------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                613 kB/s |  20 MB  00:00:32     
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Updating   : docker-ce-selinux-17.05.0.ce-1.el7.centos.noarch                                                                                  1/4 
libsemanage.semanage_commit_sandbox: Error while renaming /etc/selinux/targeted/active to /etc/selinux/targeted/previous. (Invalid cross-device link).
/usr/sbin/semodule:  Failed!
libsemanage.semanage_direct_install_info: Overriding docker module at lower priority 100 with module at priority 400.
  Updating   : docker-ce-17.05.0.ce-1.el7.centos.x86_64                                                                                          2/4 
  Cleanup    : docker-ce-17.03.1.ce-1.el7.centos.x86_64                                                                                          3/4 
  Cleanup    : docker-ce-selinux-17.03.1.ce-1.el7.centos.noarch                                                                                  4/4 
  Verifying  : docker-ce-selinux-17.05.0.ce-1.el7.centos.noarch                                                                                  1/4 
  Verifying  : docker-ce-17.05.0.ce-1.el7.centos.x86_64                                                                                          2/4 
  Verifying  : docker-ce-selinux-17.03.1.ce-1.el7.centos.noarch                                                                                  3/4 
  Verifying  : docker-ce-17.03.1.ce-1.el7.centos.x86_64                                                                                          4/4 

Updated:
  docker-ce.x86_64 0:17.05.0.ce-1.el7.centos                            docker-ce-selinux.noarch 0:17.05.0.ce-1.el7.centos                           

Complete!
`

**Comment by vmbaraiya** [comment]
Thanks @andrewhsu ...!

**Comment by tvprasad** [comment]
It did not work from me. Here the work around I used.

I used following command to mannual install by by specifying the .rpm Link

yum install https://download.docker.com/linux/centos/7/x86_64/stable/Packages/docker-ce-selinux-17.03.2.ce-1.el7.centos.noarch.rpm

That's it. It installed docker on my centos 7.

Here is the link to available docker RPM and stable packages
https://download.docker.com/linux/centos/7/x86_64/stable/Packages/

**Comment by SaadUllahSheikh** [comment]
As of today I guess Docker already fixed issue of multiple repo call using same URL 
yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
yum install docker-ce
systemctl enable docker && systemctl start docker 

I have used it and its working fine

**Comment by Akanksha1709** [comment]
>sudo yum remove docker \
                  docker-client \
                  docker-client-latest \
                  docker-common \
                  docker-latest \
                  docker-latest-logrotate \
                  docker-logrotate \
                  docker-engine
Use this to remove old version then this
yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
yum install docker-ce
systemctl enable docker && systemctl start docker
upgraded to latest version

**Comment by kingdavid72** [comment]
> It did not work from me. Here the work around I used.
> 
> I used following command to manual install by by specifying the .rpm Link
> 
> yum install https://download.docker.com/linux/centos/7/x86_64/stable/Packages/docker-ce-selinux-17.03.2.ce-1.el7.centos.noarch.rpm
> 
> That's it. It installed docker on my CentOS 7.
> 
> Here is the link to available docker RPM and stable packages
> https://download.docker.com/linux/centos/7/x86_64/stable/Packages/

also Worked for me

**Comment by kingdavid72** [comment]
> A new obsoletes restriction was introduced in docker-ce 17.06.0 and it looks like the yum repo is applying it to all versions of the docker-ce indiscriminately. To work around this when installing older versions of docker-ce, you can pass a flag to ignore obsoletes:
> 
> ```
> $ yum install -y --setopt=obsoletes=0 \
>   docker-ce-17.03.1.ce-1.el7.centos \
>   docker-ce-selinux-17.03.1.ce-1.el7.centos
> ```
> 
> Related to issue: [docker/for-linux#20 (comment)](https://github.com/docker/for-linux/issues/20#issuecomment-312122325)

sudo yum install -y --setopt=obsoletes=0 \
  docker-ce-17.03.1.ce-1.el7.centos \
  docker-ce-selinux-17.03.1.ce-1.el7.centos

Really works ^^

**Comment by bbhxwl** [comment]
me too
